### PR TITLE
Fix variation pagination after dismissing the save warning

### LINKED ISSUE #35554: Page navigation in product variations has unexpected UX behaviour
### Prerequisites

- [X] I have carried out troubleshooting steps and I believe I have found a bug.
- [X] I have searched for similar bugs in both open and closed issues and cannot find a duplicate.

### Describe the bug

In a variable product with many variants, navigating between the pages of variants will discard your changes even if you press "cancel" after the warning of unsaved changes.

### Expected behavior

When pressing "cancel", it should not navigate away, so I have a chance to save changes and not lose them.
Pressing "cancel" should abort what you're about to do, in this case navigating. It is normal that pages ask if you are sure you want to leave page.

What's happenning here is the inverse on a common behaviour; "OK" to navigate (*but it saves*), meanwhile "Cancel" navigates and throws everything away.

Either normal behaviour should occurr where it asks if you are sure you want to navigate away from unsaved changes, where "Cancel" aborts navigation and offers you a chance to use save button, or the "Cancel" label should be changed (*if possible*).

### Actual behavior

It navigates and discard all changes when I click "cancel".

### Steps to reproduce

1. Create a variable product
2. Add enough variants to have pagination pop up
3. Make changes to variants
4. Click on a navigation to another page
5. Click "Cancel" when it warns you that you are about to navigate away from unsaved changes

### WordPress Environment

WordPress 6.1
WooCommerce 7.1.0

### Isolating the problem

- [X] I have deactivated other plugins and confirmed this bug occurs when only WooCommerce plugin is active.
- [ ] This bug happens with a default WordPress theme active, or [Storefront](https://woocommerce.com/storefront/).
- [X] I can reproduce this bug consistently using the steps above.

### PR DESCRIPTION
**Pull request by vladolaru**
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

The classic variation editor currently continues to another variation page when a merchant dismisses the unsaved-changes confirmation, discarding the current edits. This change makes merchant-triggered pagination honor the dismissed confirmation, restoring the current page selector and the dirty rows.

To keep server-side mutation refreshes (save, cancel, remove, generate/link all, bulk edit) independent of that prompt, `go_to_page()` no longer navigates by simulating a merchant page change through the selector's `change` event. It calls a new shared `navigate()` helper directly, so the prompt and its abort live only on the merchant-initiated path (selector changes and the pagination arrows). This also removes a pre-existing race defect: when a mutation response arrived while the just-accepted save was still in flight, a redundant second save prompt appeared, and accepting it re-submitted the save, firing the save hooks twice. Programmatic refreshes now never prompt.

It also adds a Playwright regression covering the dialog dismissal, selected page, unsaved value, dirty state, and enabled save action on a product with enough variations to paginate, and fixes four pre-existing lint findings in the touched files in a lead-off commit.

**Backward-compatibility impact:** programmatic refreshes no longer fire the `change` event on `.variations-pagenav .page-selector`. Within core, nothing but the pagination handler listened to it; third-party code observing that event loses the signal for internal refreshes only — the `woocommerce_variations_loaded` event still fires for every page load, and external code that itself triggers `change` on the selector keeps the merchant behavior (prompt included). No PHP, hook, or signature changes.

Closes #35554.

### Screenshots or screen recordings:

Not applicable; this corrects the behavior of the existing confirmation and pagination controls without changing their appearance or copy.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://developer.woocommerce.com/docs/contribution/testing/writing-high-quality-testing-instructions/), include your detailed testing instructions:

1. Create or edit a variable product.
2. Add two attributes used for variations, with four values each, then generate all 16 variations and publish or update the product.
3. In **Product data > Variations**, stay on the first variation page, expand the first variation, and change its regular price without selecting **Save changes**.
4. Select the next variation page, then choose **Cancel** in the confirmation asking whether to save the changes.
5. Verify that the first page remains selected, the edited price is still present, and **Save changes** remains enabled.
6. Select the next page again, accept the save confirmation, and verify that the edit is saved and the second variation page loads.
7. Remove a variation, confirm the removal, and verify that the variation list refreshes and the variation count decreases.
8. Edit a regular price again without saving, select **Generate variations**, accept both confirmations, and verify that the list refreshes with no further prompts and the edited price was saved.

### Testing that has already taken place:

- Reproduced the original loss of the page, edit, and dirty state, then verified the fix in this repository's isolated local `wp-env` environment with a disposable 16-variation product.
- Passed the full `update-variations` and `create-variations` Playwright suites (11 tests) against the final build.
- Mutation-tested the guard: removing the abort `return`, the dirty-marker restoration, or the selector reset each makes the new regression fail on a distinct assertion; the earlier caller-flag design left its programmatic-navigation clause unkillable by any test, which motivated the `navigate()` split — no untestable branch remains.
- Reproduced the save-vs-mutation ajax race deterministically with a temporary route-interception spec (save response held until the generate response landed): on this branch the refresh completes with exactly two prompts and a single `woocommerce_save_variations` request; on the flag design the same spec captured a third redundant save prompt.
- Built the classic assets successfully with `pnpm wc:build:classic-assets` after every change.
- Passed Oxlint with zero findings on the changed files (the four pre-existing findings are fixed in the lead-off commit), changelog validation, and `git diff --check`.

### Milestone

- [x] Automatically assign milestone for the **[next WooCommerce version](../blob/trunk/plugins/woocommerce/woocommerce.php#L6)**

> **Note:** Check the box above to have the milestone automatically assigned when merged.
> Alternatively (e.g. for point releases), manually assign the appropriate milestone.

### Changelog entry

-   [ ] Automatically create a changelog entry from the details below.

-   [x] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

Created manually.

</details>

### Release Communication

Select if this PR needs a generated summary for release notes:

- [ ] **Feature Highlight** - For user-facing features (what changed, user impact)
- [ ] **Developer Advisory** - For developer-facing changes (what changed, how to detect, actions needed)

<details>
<summary>When to use each?</summary>

**Feature Highlight**: New features, UI changes, or improvements that merchants/store owners will notice.
- Example: "New bulk editing for products", "Improved checkout performance"

**Developer Advisory**: Breaking changes, deprecations, or changes that affect themes/plugins/extensions.
- Example: "Hook signature change", "Deprecated filter", "REST API field removed"

An AI will analyze your PR and post a draft comment for you to review and edit.
</details>

### CODE REVIEW

File: plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js [hunk 1]
```diff
@@ -1629,6 +1629,37 @@ jQuery( function ( $ ) {
...
+
+			wc_meta_boxes_product_variations_pagenav.change_classes(
+				page,
+				parseInt( wrapper.attr( 'data-total_pages' ), 10 )
+			);
+			wc_meta_boxes_product_variations_ajax
+				.load_variations( page )
+				.then(
+					wc_meta_boxes_product_variations_ajax.show_hide_variation_empty_state()
+				);
```
[the remark this answers is not in the document]
Comment (vladolaru) [reply plugins/woocommerce/client/legacy/js/admin/meta-boxes-product-variation.js:1660]: Fixed in 2e7f570417 — `navigate()` now passes the function reference, matching `reload()`.

One correction to the analysis: nothing throws here. `load_variations()` returns a native Promise, and per spec `.then()` ignores non-function arguments, so the chain resolved silently. The real defect was that the toggle ran synchronously before the request and the `.then()` held a dead handler — an idiom this refactor had carried over verbatim from the former `page_selector()` tail. The end state was coincidentally identical either way (the toggle only reads `data-total`, which `set_paginav()` updates before `navigate()` runs), but the reference form now does what the code says.

Verified with the full `update-variations` + `create-variations` Playwright suites (11 passed) on rebuilt classic assets.

**Review by elazzabi — APPROVED** [review]
The testing instructions check out, and the code looks good! ✅ 

One thing I wanted to note, so we can probably document/track it for the future (not blocking this PR): `remove_variation`, `link_all_variations`, and the `bulk-edit` path all discard `check_for_changes()`'s return value, so the same "dismiss-loses-your-edits" bug this PR fixes for pagination still exists at those three call sites.

@vladolaru can you double-check this from your end? We might need to fix it in another PR later

**Comment by vladolaru** [comment]
Hey Ahmed — good catch, and it holds up. I dug into all three and also reproduced it in wp-env rather than just reading the code, because I wanted to be sure before we filed anything.

You're right that all three discard the return value. What I hadn't appreciated until I traced it is *why* that matters so much: `check_for_changes()` is destructive on its own. When the merchant dismisses the prompt it strips `variation-needs-update` from every dirty row, leaves the edited values sitting in the inputs, and doesn't re-disable the Save/Cancel buttons — those only get re-disabled after a variations reload or a successful save. So the `false` return is really saying "abort, I already threw the dirty state away."

Where it gets worse than the pagination case: `remove_variation` and `link_all_variations` each have a **second** `window.confirm()` right after. Dismiss both and the merchant is left on the page with their edits visible, the Save button enabled, and no row marked dirty. And `save_changes()` wraps its whole body in `if ( 0 < need_update.length )` — so clicking Save does *nothing at all*. No request, no error, no feedback. That's nastier than the bug this PR fixes, which at least gave you a page change as a signal that something happened.

I confirmed exactly that in the browser: after dismissing both prompts, no `woocommerce_save_variations` request is sent to `admin-ajax.php`, and the REST API still returns the original prices. A control run — same edit, Save clicked without the remove-cancel sequence — does fire the request and does persist, so it's genuinely the dismissed prompt causing it.

`bulk-edit` is the milder one. There's no second gate (any per-action prompt sets `cancel = true` and routes around the call entirely), so it always proceeds and the response reloads the list. Edits are still silently dropped, but there's no inconsistent UI or lying Save button left behind.

One extra thing that fell out of the repro: the prompt copy is wrong at these call sites. `i18n_edited_variations` reads *"Save changes before changing page?"* — shown for a **remove** action, where no page change is involved. It was written for pagination and reused verbatim everywhere.

So: agreed this is out of scope here, and I've filed #67429 to track it. My instinct for the fix is to go at the primitive rather than each caller — have the dismissal branch stop destroying state it isn't authorised to destroy and just return `false`, then let each caller decide. That covers all the sites at once and would let this PR's `page_selector` drop its re-add-the-classes workaround. It needs a real decision per caller though (should a dismissed save prompt abort the pending delete/link/bulk edit, or not?), which is exactly why it doesn't belong in this PR.
